### PR TITLE
Adds Suffix Property for CupertinoTextFormFieldRow

### LINKED ISSUE #103385: Proposal to add suffix option in `CupertinoTextFormFieldRow`
`CupertinoTextField already have it but it did not support validation

### PR DESCRIPTION
**Pull request by piedcipher**
Adds `suffix` property for `CupertinoTextFormFieldRow`.

screenshot_2023-06-18_at_7 23 18_pm

Fixes #103385

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by piedcipher** [comment]
I've a confusion:

For `CupertinoTextFormFieldRow` widget the `prefix` is shown via `CupertinoFormRow`. `CupertinoTextField.borderless` (child of `CupertinoFormRow`) is also having a `prefix` property.

So I was wondering whether `suffix` should be added via `CupertinoFormRow` or `CupertinoTextField.borderless`?

**Comment by Renzo-Olivares** [comment]
@piedcipher It doesn't look like `CupertinoFormRow` currently has a `suffix` property only `CupertinoTextField`. I think it is okay to pipe `suffix` through to `CupertinoTextField`. It might be a little redundant to add `suffix` to `CupertinoFormRow` since the `child` widget is essentially the `suffix` according to the documentation since it is placed at the end of the row. cc @justinmc if you have any thoughts on this.

File: packages/flutter/lib/src/cupertino/text_form_field_row.dart [hunk 1]
```diff
@@ -244,6 +246,11 @@ class CupertinoTextFormFieldRow extends FormField<String> {
   /// nature of the input.
   final Widget? prefix;
 
+  /// A widget that is displayed at the end of the row.
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/cupertino/text_form_field_row.dart:249]: Both of these paragraphs seem like they are saying the same thing. Maybe replace with `An optional [Widget] that is displayed at the end of the row.`

**Review by Renzo-Olivares — CHANGES_REQUESTED** [review]

File: packages/flutter/lib/src/cupertino/text_form_field_row.dart (hunk 1 above)
Comment (piedcipher) [reply packages/flutter/lib/src/cupertino/text_form_field_row.dart:249]: Sure

Comment (piedcipher) [reply packages/flutter/lib/src/cupertino/text_form_field_row.dart:249]: Done 👍✅

File: packages/flutter/lib/src/cupertino/text_form_field_row.dart [hunk 2]
```diff
@@ -232,6 +233,7 @@ class CupertinoTextFormFieldRow extends FormField<String> {
                 placeholder: placeholder,
                 placeholderStyle: placeholderStyle,
                 contextMenuBuilder: contextMenuBuilder,
+                suffix: suffix,
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/cupertino/text_form_field_row.dart:236]: I think we should also expose `suffixMode` so the user can control the visibility.

Comment (piedcipher) [reply packages/flutter/lib/src/cupertino/text_form_field_row.dart:236]: Sure, I'll make the necessary changes.

**Review by justinmc — COMMENTED** [review]
I agree about adding `suffixMode`, otherwise this looks good.

 > So I was wondering whether suffix should be added via CupertinoFormRow or CupertinoTextField.borderless?

I'm not sure about the history of `child` vs. `suffix` on CupertinoFormRow.  If putting it on CupertinoTextField.borderless looks good then I'm on board with that.  The code seems straightforward.

CC @MitchellGoodwin in case you happen to have an opinion on that.

**Comment by MitchellGoodwin** [comment]
For CupertinoFormRow, it looks like it was added with several other widgets, so I wonder if there just wasn't a conversation of `child` vs. `suffix`. My opinion is that ideally CupertinoFormRow's `child` property would already be called `suffix`, as since it has two children widgets, why would one be the child widget and the other be `prefix`?

Maybe a `suffix` property should be added to CupertinoFormRow and `child` would be deprecated for consistency's sake? But I'm not sure if it's worth forcing the migration. I do not think a `suffix` property should be added otherwise, when `child` could be used.

**Comment by justinmc** [comment]
@piedcipher Do you still have plans to come back to this PR?  Need help with anything?

**Comment by piedcipher** [comment]
> @piedcipher Do you still have plans to come back to this PR? Need help with anything?

Sorry, I got caught up at work hence couldn't update my pr. But now I've got some time. I'll resume my work on this pr from today.

**Comment by piedcipher** [comment]
> For CupertinoFormRow, it looks like it was added with several other widgets, so I wonder if there just wasn't a conversation of `child` vs. `suffix`. My opinion is that ideally CupertinoFormRow's `child` property would already be called `suffix`, as since it has two children widgets, why would one be the child widget and the other be `prefix`?
> 
> Maybe a `suffix` property should be added to CupertinoFormRow and `child` would be deprecated for consistency's sake? But I'm not sure if it's worth forcing the migration. I do not think a `suffix` property should be added otherwise, when `child` could be used.

Hi folks! I've a doubt. CupertinoFormRow already have a child and if we're not planning a migration how to add Suffix. I'm confused about that.

**Comment by piedcipher** [comment]
> @piedcipher Do you still have plans to come back to this PR? Need help with anything?

Finally I've got some time now. Sorry about the delay. I'm confused regarding how to add Suffix since child is non-nullable in CupertinoFormRow.

Also if somehow we add a suffix to CupertinoFormRow, should we remove it from CupertinoTextFormFieldRow?

**Comment by justinmc** [comment]
@piedcipher I  see you pushed some commits here, let me know if/when you need a re-review. No rush.

**Comment by piedcipher** [comment]
> @piedcipher I see you pushed some commits here, let me know if/when you need a re-review. No rush.

@justinmc i only resolved merge conflicts, no new changes.

I'm still stuck on https://github.com/flutter/flutter/pull/129077#issuecomment-1705639611, i forgot to revisit this pr.

**Comment by goderbauer** [comment]
@piedcipher Do you still have plans to get back to this PR and resolve the feedback given above?

**Comment by piedcipher** [comment]
> @piedcipher Do you still have plans to get back to this PR and resolve the feedback given above?

Hey still in dilemma https://github.com/flutter/flutter/pull/129077#issuecomment-1705639611

But I think we can close this someone can take this to finish line.

**Pull request closed without merging**